### ISSUE TITLE
weird results

### ISSUE THREAD

**Issue by mhusseinsh** [junyanz/pytorch-CycleGAN-and-pix2pix#336]
Hello,

I am training the CycleGAN on driving scenes, in the beginning the results were a little bit nicer but then there were a little bit weird

This is after epoch 18
![image](https://user-images.githubusercontent.com/33177438/43121548-6fe38350-8f1e-11e8-80e7-0beaf0c44dd2.png)

This is after epoch 92
![image](https://user-images.githubusercontent.com/33177438/43121573-865be154-8f1e-11e8-98c0-0a62f93cdd36.png)
![image](https://user-images.githubusercontent.com/33177438/43121590-8dd7dff0-8f1e-11e8-9922-2e4be47a6828.png)

During the training, the results generated and saved in the checkpoints are much better and well translated.
However, when i load the latest checkpoint on test images, these results up there I have @junyanz

any idea why this is happening ?

**Comment by mhusseinsh** [comment]
During the training, the results generated and saved in the checkpoints are much better and well translated.
However, when i load the latest checkpoint on test images, these results up there I have @junyanz

**Comment by junyanz** [comment]
Could you share with your training and test script? 
Could you also try to run your saved model on the training set images and see if it can match the saved results?

**Comment by mhusseinsh** [comment]
@junyanz  This is how I run and test
`python train.py --dataroot ./datasets/kitti--name kitti_cyclegan --model cycle_gan --gpu_ids=6 --display_id -1`

`python test.py --dataroot datasets/kitti/testA --name kitti_cyclegan --checkpoints_dir ./checkpoints/ --model test --gpu_ids=7 --resize_or_crop none`

**Comment by junyanz** [comment]
Interesting. Does it work without `--resize_or_crop none`? 

how about running a test with cyclegan model directly?
```bash
python test.py --dataroot ./datasets/kitti--name kitti_cyclegan --model cycle_gan --gpu_ids=6
```

**Comment by taesung89** [comment]
I think the problem might be that you scaled the image to 256px x 256px at training time, but at test time, you used the 800px x 600px original resolution. This is a pretty big gap. 

I recommend you first test without --resize_or_crop none to see this is the real problem. 
Then I recommend training and test at the same scale. You can do this by 

`--resize_or_crop crop --fineSize 360`

which loads the image at the original resolution of 800x600 and then making a square crop of 360x360. Please change the number 360 to something that fits on your GPU. 

This method does not change the scale of the image, so you can use `--resize_or_crop none` option at test time.

**Comment by mhusseinsh** [comment]
**how about running a test with cyclegan model directly?**
@junyanz same results with cyclegan model directly

**I recommend you first test without --resize_or_crop none to see this is the real problem.**
@taesung89 also same problem
![image](https://user-images.githubusercontent.com/33177438/43253543-886f784a-90c5-11e8-9e0f-1f916579ddd2.png)

**Comment by mhusseinsh** [comment]
@taesung89 maybe it is a little bit better
`--resize_or_crop none`
![image](https://user-images.githubusercontent.com/33177438/43253826-28db1136-90c6-11e8-8658-2b44eefc9d93.png)

without `--resize_or_crop none`
![image](https://user-images.githubusercontent.com/33177438/43253846-30a164b0-90c6-11e8-8989-b94319f4bc9e.png)


But this is for an image in the test set, in general, it is not performing good on the test set compared to what was saved during training

**Comment by junyanz** [comment]
There is always a training/test gap in any ML system. The key is to reproduce the training set results with the test script.

**Comment by mhusseinsh** [comment]
@junyanz  Sorry I don't get your point. What do you mean by reproducing the training set results with the test script ?

**Comment by junyanz** [comment]
You can run the model on the training images and see if it is the same as the saved results during training.

**Comment by mhusseinsh** [comment]
@junyanz yes this is what I did, as in my previous comments

**Comment by duburlan** [comment]
@mhusseinsh at test time, try to set `model.train()`. I think it is due to instance normalization.

**Comment by junyanz** [comment]
The results look like the saved one without `--resize_or_crop none`？

**Comment by mhusseinsh** [comment]
@junyanz which one do you mean ?

**Comment by junyanz** [comment]
Sorry. I was traveling for the past two weeks. You can produce results with the test script on the training images, and see if the results are the same as your "saved" training image results. 

There is often a gap between training and test due to overfitting. So it's quite common that the test images look worse compared to the training images. But to make sure that your test script is correct, you can do a sanity check using training images as described above.

**Comment by happsky** [comment]
@junyanz @taesung89 @SsnL I got the same problem when I trian pix2pix on cross-view image translation task.
During training time, the results are quite good as I wanted,
![image](https://user-images.githubusercontent.com/5948971/45509752-06946f00-b75e-11e8-9b57-85a8ce5d231b.png)
However, when I use the same images for testing and I got very bad results,
![image](https://user-images.githubusercontent.com/5948971/45509804-275cc480-b75e-11e8-9ee3-e89468c0e480.png)


My commads are:
python train.py --dataroot ./data --name setting_1 --model pix2pix --which_model_netG unet_256 --which_direction AtoB --dataset_mode aligned --norm batch --pool_size 50 --gpu_ids 0 --batch 32 --loadSize 286 --fineSize 256;


python test.py --dataroot ./data  --name setting_1 --model pix2pix --which_model_netG unet_256 --which_direction AtoB --dataset_mode aligned --norm batch --gpu_ids 0 --batchSize 32 --loadSize 256 --fineSize 256;

Any suggestion.

**Comment by junyanz** [comment]
@happsky  maybe you also want to set the loadSize=286 during test time. I also think that there is severe overfitting during training. This is an ill-posed problem. But you training set results look identical to the ground truth images.

**Comment by happsky** [comment]
@junyanz I got the similar bad results after setting loadSize=286 during test time. For avoiding overfitting during training time, do you have any suggestions?

**Comment by junyanz** [comment]
To avoid overfitting: (1) increase training set (2) dropout? (3) more data argumentation. 
We haven't used a big batchSize before. We use to use batchSize=1. 
You can try to make the model as eval mode. Call this [function](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/models/base_model.py#L47) in the test code.

**Comment by SsnL** [comment]
@happsky @mhusseinsh This should be fixed in https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/commit/7dfdd06d8f7ca41735c06ea67ffbebd222a4d65e ! Because the pix2pix model uses batch norm, if we don't set it to eval mode, the running stats are not used, and result will look quite bad because batch size is 1 in test time. Sorry about it. Could you try to pull the repo and test again? There is no need to re-train. Just running test.py again should be fine.

**Comment by happsky** [comment]
@SsnL Thank you so much and I got quite better results now!

**Comment by happsky** [comment]
@junyanz @SsnL As I posted before,
![image](https://user-images.githubusercontent.com/5948971/45568889-87696e80-b823-11e8-8d05-8fba93c73178.png)
Why the fake_B and the real_B flipped left and right?

**Comment by junyanz** [comment]
There is random flipping in the current data augmentation. You can add `no_flip`. See [here](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/options/base_options.py#L40) for more details.

**Comment by happsky** [comment]
@junyanz Thank you for your quick response!

**Comment by junyanz** [comment]
For the eval mode, I added a [flag](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/options/test_options.py#L12) that allows you to use eval mode. In the original pix2pix paper (@phillipi ), we don't use eval mode during the test, as we often use batchSize=1, and we would like to get per-image statistics. We often get better results without eval mode. Here is a comparison of label-> facades with and without eval mode. 

No eval mode
![no_eval](https://user-images.githubusercontent.com/1924757/45574442-0c618180-b83e-11e8-837a-85ed7cebd655.png)

Eval mode
![eval](https://user-images.githubusercontent.com/1924757/45574448-11becc00-b83e-11e8-98d6-ba877ff15696.png)


But in your case, you have a big batchSize during training (32) and you use a small batchSize(=1) during test (Note: we hard-coded the batchSize=1 in our test [code](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/test.py). I will try to relax it later). 
In general, I will recommend that users use instance norm for both pix2pix and CycleGAN, which guarantees training/test behavior and also get per-image statistics.

**Comment by sahilsharma884** [comment]
For this research paper, [unpaired image translation](https://arxiv.org/pdf/1703.10593.pdf) 

![Capture](https://user-images.githubusercontent.com/49679122/83407192-cdb69500-a42d-11ea-96aa-4c186ee2d620.JPG)

Hello Sir,  I am trying with dataset **cezanne2photo** ([avaiable](https://people.eecs.berkeley.edu/~taesung_park/CycleGAN/datasets/)) with image resizing (128x128) from 256x256 (due to limit resource available). It seems that its discriminator gets overfitting. So questions based on the snapshot i have included.
 In generative model
1. Can you specify in which layer should I used dropout layer with dropout rate value?
2. Did you apply reflect padding with in the all 6 resnet blocks only or in a whole network (rather than zero padding)?
3. at the last layer, c7s1-3, did you apply padding or any activation (in actual it is ReLU)? if yes please specify? If not, then the image has to be normalized rather than standardization as ReLU activation will neglected negative values when perform standardization.

In discriminative model,
1. what was the kernel size, padding, stride and activation in the last layer after c512?

I mean after I have done several hyperparameter and testing, its always overfitting.Its was so frustrating. So, kindly can you show the actual architecture with more specific parameters (not from code). I also follow CycleGAN present in tensorflow about data augmentation.
Still resulting are not satisfying.

Thanks

**Comment by junyanz** [comment]
You can change the network according to your application. Here is the [model](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/5fd11c27a1b9a5d79ef4ab879146c6b065f6e2a8/models/networks.py#L315). 
1. No. The current dropout flag is applied to many layers. 
2. The entire network. I recommend reflection padding. 
3. The last year has a reflection padding layer according to this [line](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/5fd11c27a1b9a5d79ef4ab879146c6b065f6e2a8/models/networks.py#L365). 

For overfitting, you can increase the `--lambda_identity`. It partially alleviated the issue.
